src/client/dfuse/inval.c

```diff
@@ -0,0 +1,438 @@
+ * of the correct list.
+ *
+ * Separately have a thread which periodically will walk each list starting at the front and
+ * invalidate any entries where the timeout has expired.
+ *
+ * In this way the lists are never traversed, on access a entry is removed from where it is and
+ * appended to the end, and the timeout starts at the front of the list and traverses only as far
+ * as it needs to until the front entry is to be kept.
+ *
+ * As lookups will not be repeated by the kernel until after timeout has expired allow some leeway
+ * before eviction to allow re-validation of in-use datasets without triggering entire tree
+ * invalidations through the kernel.  Directories get five seconds, anything else two.  Ideally
+ * directories would be invalidated first as this would result in less dfuse->kernel calls as once
+ * the kernel invalidates directories then it invalidates the whole tree below that, however there
+ * are also use-cases where there are significiant numbers of files per directory where the
+ * directory is in active use but individual files are not.
+ *
+ * Future work might be to speculatively perform lookups close to the end of the timeout period,
+ * then if a entry was in frequent use it's lookup could be performed from memory, effectively
+ * moving the re-validation cost off the critical path.  This code currently only handles dentries
+ * but could also separately track attributes (inodes) and file contents as well.
+ *
+ * Additional changes to consider in the future could include:
+ *  Better handing of eviction timeouts, "max(time * 1.1, 10)" would be better than a flat +x/+5
+ *  Use arrays rather than lists for the buckets for faster iteration.
+ *  Reference counting the timeout buckets.
+ *
+ * Locking: The ival_lock is contended, it is accessed several places, however none do any more
+ * than list management.  As inodes might be removed from one list and re-inserted into another
+ * there is a pre subsystem lock rather than per list locks.
+ *  ie_close() which is called from forget and some failure paths in readdir()
+ *  lookup() to move entries to the end of this list.
+ *  de_run() to pull items from the front of the list.
+ *
+ * Wakeup: The invalidation thread is woken up when:
+ *  dfuse is exiting.
+ *  something is added to an empty list.
+ *  after a timeout.
+ * Timeouts are chosen based on the entries still on any list, dfuse will sleep as long as it can
+ * but at least 2 seconds and at most 60.
+ * As this relates to releasing resources there is no additional benefit in finer grained time
+ * control than this.
+ */
+
+/* Grace period before invalidating directories or non-directories.  Needs to be long enough so that
+ * entries in the working set are invalidated but short enough to be meaningful.
+ * Directories that are used as the cwd for processes can cause problems with being invalidated too
+ * early so use a higher value here.
+ */
+#define INVAL_DIRECTORY_GRACE (60 * 30)
```

Comment:
There's a wider problem that processes can call stat() on the inode of their current working directory but they do not hold a reference on it so if this inode expires and is invalidated then processes can start seeing errors - I'm hoping this will be fixed in later kernels.  See the associated ticket for more details.